### PR TITLE
New Rule Proj0001: .NET project file could not be located

### PR DESCRIPTION
**Pull request by Corniel**
So that other rules don't have to crash (or all report) that the .NET project file could not be resolved, this rule handles that scenario for them.

### CODE REVIEW

File: src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+public sealed class PackageReference : Node
+{
+    public PackageReference(XElement element) : base(element)
+    {
+    }
+
+    public string? Include => GetAttribute();
+
+    public string? Version => GetAttribute();
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs:13]: Is it intentional that `Include` and `Version` are the same?

Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs:13]:
``` C#
  public string? GetAttribute([CallerMemberName] string? propertyName = null)
        => Element.Attribute(propertyName)?.Value;
 ```
 
So the return the value of different attributes.

Comment (CptWesley) [reply src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs:13]: In that case I think the properties shouldn't be nullable.

File: rules/Proj0001.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Proj0001: .NET project file could not be located
+Roslyn officially does not support the analysis on .NET project files. Therefor
```
Comment (LauraXiulan) [inline rules/Proj0001.md:2]: Therefore

File: rules/Proj0001.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+all rules in this package try to resolve the project file on disk based on
```
Comment (LauraXiulan) [inline rules/Proj0001.md:3]: This sentence reads a bit weird with the two times 'on'.

File: src/DotNetProjectFile.Analyzers/tools/uninstall.ps1 [hunk 4]
```diff
@@ -0,0 +1,10 @@
+param($installPath, $toolsPath, $package, $project)
+
+$analyzersPath = Split-Path -Path $toolsPath -Parent
+$analyzersPath = Join-Path $toolsPath "analyzers"
+$analyzerFilePath = Join-Path $analyzersPath "DotNetProjectFile.Analyzers.dll"
+try {
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/tools/uninstall.ps1:6]: Why the try catch if you don't do anything with the catch?

File: src/DotNetProjectFile.Analyzers/Xml/Project.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        if (compilation.SourceModule.ContainingAssembly is { } assembly)
+        {
+            var directories = assembly.Locations.Select(l => l.SourceTree?.FilePath)
+                .Where(p => p is { })
+                .Select(p => new FileInfo(p).Directory)
+                .Distinct(DirectoryEqualityComparer.Instance)
+                .Where(f => f.Exists)
+                .ToArray();
+
+            var project = directories.Select(d => GetFile(d, $"{assembly.Name}.??proj"))
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Xml/Project.cs:29]: What does this ??proj do?

File: src/DotNetProjectFile.Analyzers/Rule.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿namespace DotNetProjectFile;
+
+public static class Rule
+{
+    public static DiagnosticDescriptor ProjectFileCouldNotBeLocated => New(
+        id: 0001,
+        title: ".NET project file could not be located",
+        message: "The project file '{0}' could not be located.",
+        description: "In order to make this rules work, the project file should be located.",
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Rule.cs:9]: these rules

File: src/DotNetProjectFile.Analyzers/Diagnostics/Category.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿namespace DotNetProjectFile.Diagnostics;
+
+public enum Category
+{
+    Clairity,
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Diagnostics/Category.cs:5]: Clarity

File: specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using CodeAnalysis.TestTools.Contexts;
+using Microsoft.CodeAnalysis;
+
+namespace FluentAssertions;
+
+internal static class FindsExtensions
+{
+    public static void HasNoIssues(this ProjectAnalyzerVerifyContext context)
+        => context.HasIssues();
```
Comment (LauraXiulan) [inline specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs:9]: This is a bit confusing, it has no issues when it has issues without parameter?

File: specs/DotNetProjectFile.Analyzers.Specs/Extensions/Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+using CodeAnalysis.TestTools.Contexts;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal static class DiagnosticAnalyzerExtensions
+{
+    public static ProjectAnalyzerVerifyContext ForProject(this DiagnosticAnalyzer analyzer, string fileName)
+    {
+        var name = Path.GetFileNameWithoutExtension(fileName);
+        var file = new FileInfo($"../../../../../projects/{name}/{fileName}proj");
```
Comment (LauraXiulan) [inline specs/DotNetProjectFile.Analyzers.Specs/Extensions/Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer.cs:11]: Does this mean that the fileName always has the .cs/.vb extension? What happens if it is called with another extension?

File: src/DotNetProjectFile.Analyzers/tools/uninstall.ps1 (hunk 4 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/tools/uninstall.ps1:6]: This is what Microsoft suggests. So I just followed that:

https://learn.microsoft.com/en-us/nuget/guides/analyzers-conventions#install-and-uninstall-scripts

File: rules/Proj0001.md (hunk 3 above)
Comment (Corniel) [reply rules/Proj0001.md:3]: I dropped "on disk"

File: src/DotNetProjectFile.Analyzers/Xml/Project.cs (hunk 5 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Xml/Project.cs:29]: Matching both .csproj (C#) as vbproj (VB.NET). It is a bit lazy, so I could be improved on.

File: specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs (hunk 8 above)
Comment (Corniel) [reply specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs:9]: `HasIssues` has `param Issue[] issues`, so without params is equivalent to no issues.

Comment (Corniel) [reply specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs:9]: I've tried to make that easier to grasp.

File: src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs (hunk 1 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Xml/PackageReference.cs:13]: What do you mean? Null means: attribute or element does not exist.

**Pull request merged**